### ISSUE TITLE
False positive (and puzzling message) with union types and callables

### ISSUE THREAD

**Issue by clo-vis** [python/mypy#13720]
**To Reproduce**

This works (no complains from mypy):
```python
from typing import TypeVar, Callable

_T = TypeVar("_T")

class Param:
    def __init__(self, initialize: Callable[[], _T]) -> None:
        ...

Param(lambda: 1)
```

as well as:
```python
from typing import TypeVar, Callable

_T = TypeVar("_T")

class Param:
    def __init__(self, initialize: _T | Callable[[], _T]) -> None:
        ...

Param(1)
```

but for this:
```python
from typing import TypeVar, Callable

_T = TypeVar("_T")

class Param:
    def __init__(self, initialize: _T | Callable[[], _T]) -> None:
        ...

Param(lambda: 1)
```

mypy issues two weird message:
```
src\playground.py:9: error: Argument 1 to "Param" has incompatible type "Callable[[], int]"; expected "Callable[[], <nothing>]"
src\playground.py:9: error: Return statement in function which does not return
Found 2 errors in 1 file (checked 1 source file)

```
